### PR TITLE
[core] Large/full refactor of `from_pretrained`

### PR DESCRIPTION
**Pull request by Cyrilvallez**
# What does this PR do?

## Preview

Large refactor of `from_pretrained`. The most important updates are the following:

- Much more readable and maintainable in the future (hopefully, at least IMO). Simplified a looooot of weird/dead/useless code that accumulated over the years
- faster model downloads (concurrent file download)
- new keyword argument `key_mapping`, allowing direct mapping of the weight names if loading a model from the hub which is compatible to given arch, but was not converted accordingly. For example, the following snippet works nicely:
```py
from transformers import LlamaForCausalLM, AutoTokenizer, LlamaConfig

key_mapping = {
    r"^transformer.wte": r"model.embed_tokens",
    r"^transformer.rotary": r"model.rotary_emb", 
    r"^transformer.ln_f": r"model.norm", 

    r"^transformer.h.(\d+).ln_1": r"model.layers.\1.input_layernorm",
    r"^transformer.h.(\d+).ln_2": r"model.layers.\1.post_attention_layernorm",

    r"^transformer.h.(\d+).mlp.c_fc_0": r"model.layers.\1.mlp.gate_proj",
    r"^transformer.h.(\d+).mlp.c_fc_1": r"model.layers.\1.mlp.up_proj",
    r"^transformer.h.(\d+).mlp.c_proj": r"model.layers.\1.mlp.down_proj",

    r"^transformer.h.(\d+).attn.attention.k_proj": r"model.layers.\1.self_attn.k_proj",
    r"^transformer.h.(\d+).attn.attention.v_proj": r"model.layers.\1.self_attn.v_proj",
    r"^transformer.h.(\d+).attn.attention.q_proj": r"model.layers.\1.self_attn.q_proj",
    r"^transformer.h.(\d+).attn.attention.out_proj": r"model.layers.\1.self_attn.o_proj",
}

model_id = "LGAI-EXAONE/EXAONE-3.0-7.8B-Instruct"

config = LlamaConfig(num_hidden_layers=32,
                     num_key_value_heads=8,
                     intermediate_size=14336,
                     hidden_size=4096,
                     bos_token_id=1,
                     eos_token_id= 361,
                     pad_token_id=0,
                     rope_theta=500000.0,
                     vocab_size=102400
                     )
model = LlamaForCausalLM.from_pretrained(model_id, config=config, key_mapping=key_mapping, torch_dtype="float16", device_map=0)
```
and will allow to simplify model conversions. It will also help when teams want to add their models to the library, but tey are 1:1 compatible with existing archs.

## Some pointers

Files modified apart from `modeling_utils.py` and `hub.py` are due to the removal of legacy functions `get_file_from_repo` and `_load_pretrained_model_low_mem` during the cleanup.

Note that I was **as careful as can be** during the refactor since this is extremely critical code. All tests are passing on the CI, and I tested the most common scenarios, as well as the new features added.  Everything seem to work.

However, I am deeply sorry for the person reviewing @ArthurZucker 😆🥲🤗

### CODE REVIEW

**Comment by ydshieh** [comment]
Hi @Cyrilvallez 

The workflow run: https://github.com/huggingface/transformers/actions/runs/13249941627
The slack report: https://huggingface.slack.com/archives/C06LR9PQA00/p1739225850225799
The new failures: https://huggingface.co/datasets/hf-internal-testing/transformers_daily_ci/blob/13935d4e1231eb06ab3ce9647a6d3ec92a85b16e/2025-02-10/ci_results_run_models_gpu/new_model_failures.json

**Comment by Cyrilvallez** [comment]
All our test suite passes. See [here](https://huggingface.co/datasets/hf-internal-testing/transformers_daily_ci/blob/495f87e77baa0078f7290adc88fb002b7df4ac52/2025-02-12/ci_results_run_models_gpu/new_model_failures.json) for new failures compared to last full CI run, but I personally checked that they are all unrelated and are due to the diff between the tip of this branch and `main`

File: src/transformers/utils/hub.py [hunk 1]
```diff
@@ -432,13 +434,9 @@ def cached_file(
     return resolved_file
 
 
-# TODO: deprecate `get_file_from_repo` or document it differently?
-#       Docstring is exactly the same as `cached_repo` but behavior is slightly different. If file is missing or if
-#       there is a connection error, `cached_repo` will return None while `get_file_from_repo` will raise an error.
-#       IMO we should keep only 1 method and have a single `raise_error` argument (to be discussed).
-def get_file_from_repo(
-    path_or_repo: Union[str, os.PathLike],
-    filename: str,
```
Comment (SunMarc) [inline src/transformers/utils/hub.py:441]: Let's not deprecate this right now and wait a few version before removing it + add a deprecation msg indicating that they can use `cached_file` instead.

File: src/transformers/utils/hub.py [hunk 2]
```diff
@@ -432,13 +434,9 @@ def cached_file(
...
+def cached_files(
+    path_or_repo_id: Union[str, os.PathLike],
+    filenames: List[str],
```
Comment (SunMarc) [inline src/transformers/utils/hub.py:439]: is there a way to make use of the `cached_file` function ?

File: src/transformers/utils/hub.py [hunk 3]
```diff
@@ -1045,44 +1160,34 @@ def get_checkpoint_shard_files(
...
-                force_download=force_download,
-                proxies=proxies,
-                resume_download=resume_download,
-                local_files_only=local_files_only,
-                token=token,
-                user_agent=user_agent,
-                revision=revision,
-                subfolder=subfolder,
-                _commit_hash=_commit_hash,
-            )
```
Comment (SunMarc) [inline src/transformers/utils/hub.py:1071]: This is pretty short already compared to the implementation of `caches_files` + we don't have the tqdm bar anymore with your changes

**Review by SunMarc — COMMENTED** [review]
Thanks for this huge work @Cyrilvallez ! Left a few comments

**Comment by poedator** [comment]
Hi, @Cyrilvallez! This refactor was long needed. Thank you for streamlining `modeling_utils.py`!

is it possible to avoid creating `state_dict` in RAM altogether and loading model params straight to GPU(s)? sometimes a server has too little RAM or it may be taken by `tmpfs`.
Perhaps single param/buffers items could be sent to GPU right after loading to disk, without waiting for  whole `state_dict` to load?

Also please ensure that quantized models loading/saving is tested in this PR.

**Comment by ydshieh** [comment]
> Also please ensure that quantized models loading/saving is tested in this PR.

Hmm, indeed, I need to trigger quantization CI too!

**Comment by Cyrilvallez** [comment]
Oh I thought we did already! No worries, I'll ping you on slack @ydshieh, I first have to do a few changes.

@poedator thanks! Concerning the loading, note that state dicts on the hub are usually quite small (never bigger than 10 GB), so loading everything in ram is never an issue. It would otherwise probably be very slow!

**Comment by ydshieh** [comment]
> Oh I thought we did already!

My bad. I am (almost always) focus on the CI jobs regarding `tests/models/xxx`, but this time it is indeed important to check the CI jobs regarding `quantization/xxx`

**Comment by Cyrilvallez** [comment]
Every state dicts are loaded one after the other, so no need to worry! See e.g. [here](https://huggingface.co/meta-llama/Llama-3.3-70B-Instruct/tree/main) for Llama 70B: each state dict is at most 5 GB

File: src/transformers/modeling_utils.py [hunk 4]
```diff
@@ -924,6 +897,674 @@ def _add_variant(weights_name: str, variant: Optional[str] = None) -> str:
...
+    # If a module has the same name under the base and task specific model, we have to re-add it to unexpected keys (it
+    # means that it is present 2 times in the adjusted loaded keys)
+    if loading_base_model_from_task_state_dict:
+        for key in set(loaded_keys):
+            if sum(key == x for x in loaded_keys) > 1:
+                unexpected_keys.add(key)
+
+    # Remove nonpersistent buffers from unexpected keys: they are not in the expected keys (model state dict), but
+    # may be in the loaded keys. Note that removing all buffers does the job, as they were part of the expected keys anyway
+    model_buffers = {n for n, _ in model.named_buffers()}
```
[the remark this answers is not in the document]
Comment (Cyrilvallez) [reply src/transformers/modeling_utils.py:1410]: Yes, but sometimes they still exist in old checkpoints on the Hub, so we don't want to raise unnecessary warnings 😬

File: src/transformers/modeling_utils.py [hunk 5]
```diff
@@ -924,6 +897,674 @@ def _add_variant(weights_name: str, variant: Optional[str] = None) -> str:
...
+    unexpected_keys = sorted(unexpected_keys - model_buffers)
+
+    # Old checkpoints may have keys for rotary_emb.inv_freq for each layer, however we moved this buffer to the main model
+    # (so the buffer name has changed). Remove them in such a case
+    has_inv_freq_buffers = any(buffer.endswith("rotary_emb.inv_freq") for buffer in model_buffers)
+    if has_inv_freq_buffers:
+        unexpected_keys = [k for k in unexpected_keys if "rotary_emb.inv_freq" not in k]
```
[the remark this answers is not in the document]
Comment (Cyrilvallez) [reply src/transformers/modeling_utils.py:1417]: Similar answer as above... I think the buffers used to be persistent at some point, then were marked as non-persistent. But they remain on all checkpoints unfortunately...

File: src/transformers/modeling_utils.py [hunk 6]
```diff
@@ -4498,456 +4714,312 @@ def _fix_state_dict_keys_on_save(self, state_dict):
...
+        start_prefix_to_remove = ""
+        # In this case, we load a BaseModel from a ForTaskModel state dict
+        if loading_base_model_from_task_state_dict:
+            start_prefix_to_remove = _prefix
+        # In this case, we load a ForTaskModel with keys from a BaseModel -> only load keys to the BaseModel
+        elif loading_task_model_from_base_state_dict:
+            model_to_load = getattr(model, prefix)
+            # We need tp update the device map as well
+            if device_map is not None:
+                device_map = {k.replace(start_prefix_to_remove, ""): v for k, v in device_map.items()}
```
[the remark this answers is not in the document]
Comment (Cyrilvallez) [reply src/transformers/modeling_utils.py:4766]: nice catch here!

File: src/transformers/modeling_utils.py [hunk 7]
```diff
@@ -4498,456 +4714,312 @@ def _fix_state_dict_keys_on_save(self, state_dict):
...
+            loaded_state_dict_keys = list(state_dict.keys())
+        else:
+            loaded_state_dict_keys = list(load_state_dict(checkpoint_files[0], weights_only=weights_only).keys())
 
-        # tie the model weights before retrieving the state_dict
-        model.tie_weights()
+        # Rename the keys (use dummy values in input dict as a small trick)
+        renamed_loaded_keys = list(
+            cls._fix_state_dict_keys_on_load({key: "" for key in loaded_state_dict_keys}, key_mapping).keys()
+        )
```
[the remark this answers is not in the document]
Comment (Cyrilvallez) [reply src/transformers/modeling_utils.py:4739]: Renamed everything as `all_checkpoint_keys`: it's a bit more informative IMO

File: src/transformers/modeling_utils.py [hunk 8]
```diff
@@ -4498,456 +4714,312 @@ def _fix_state_dict_keys_on_save(self, state_dict):
...
-                    loaded_keys,
-                    original_loaded_keys,
-                    add_prefix_to_model,
-                    remove_prefix_from_model,
-                    ignore_mismatched_sizes,
+            # Modify the keys if needed
+            state_dict = cls._fix_state_dict_keys_on_load(state_dict, key_mapping)
+            # Remove the prefix if needed
+            if loading_base_model_from_task_state_dict:
+                state_dict = _fix_state_dict_prefix(state_dict, start_prefix_to_remove)
```
[the remark this answers is not in the document]
Comment (Cyrilvallez) [reply src/transformers/modeling_utils.py:4919]: IMO the `if` really helps to quickly understand what is happening, instead of applying a function which does nothing most of the time 🙂

File: src/transformers/utils/hub.py (hunk 3 above)
Comment (Cyrilvallez) [reply src/transformers/utils/hub.py:1050]: The bar was removed as `cached_files` already displays different bars (one for each file being downloaded) through the `snapshot_download` function 🤗

File: src/transformers/utils/hub.py (hunk 2 above)
Comment (Cyrilvallez) [reply src/transformers/utils/hub.py:417]: No, but the other way around should be possible! Just refactored `cached_file` to call `cached_files`

File: src/transformers/modeling_utils.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (SunMarc) [reply src/transformers/modeling_utils.py:4911]: Sounds good !

File: src/transformers/modeling_utils.py [hunk 9]
```diff
@@ -4585,38 +4569,45 @@ def _fix_state_dict_key_on_load(key) -> Tuple[str, bool]:
...
+            new_key, has_changed = self._fix_state_dict_key_on_load(key)
+
+            # Optionally map the key according to `key_mapping`
+            if key_mapping is not None:
+                for pattern, replacement in key_mapping.items():
+                    new_key, n_replace = re.subn(pattern, replacement, new_key)
+                    # Early exit of the loop
+                    if n_replace > 0:
+                        has_changed = True
+                        break
```
Comment (ArthurZucker) [inline src/transformers/modeling_utils.py:4598]: can be done without a for loop but no worries

File: src/transformers/modeling_utils.py [hunk 10]
```diff
@@ -4585,38 +4569,45 @@ def _fix_state_dict_key_on_load(key) -> Tuple[str, bool]:
...
-
-    def _fix_state_dict_keys_on_load(self, state_dict):
+    def _get_key_renaming_mapping(
+        self,
+        checkpoint_keys: List[str],
+        key_mapping: Optional[Dict[str, str]] = None,
+        loading_base_model_from_task_state_dict: bool = False,
+        loading_task_model_from_base_state_dict: bool = False,
+    ):
         """Fixes state dict keys by replacing legacy parameter names with their modern equivalents.
```
Comment (ArthurZucker) [inline src/transformers/modeling_utils.py:4579]: updaet as this is the single point of change for key names etc

File: src/transformers/modeling_utils.py [hunk 11]
```diff
@@ -4653,441 +4644,307 @@ def _fix_state_dict_keys_on_save(self, state_dict):
...
-                )
-                with deepspeed.zero.GatheredParameters(not_initialized_parameters, modifier_rank=0):
-                    model.apply(model._initialize_weights)
-            else:
-                model.apply(model._initialize_weights)
-
-        # Set some modules to fp32 if any
-        if keep_in_fp32_modules == []:
-            keep_in_fp32_modules = None
+            model._move_missing_keys_from_meta_to_cpu(missing_keys, unexpected_keys, dtype, hf_quantizer)
```
Comment (ArthurZucker) [inline src/transformers/modeling_utils.py:4709]: noice!

File: src/transformers/modeling_utils.py [hunk 12]
```diff
@@ -5894,26 +5786,24 @@ def caching_allocator_warmup(model: PreTrainedModel, expanded_device_map: Dict,
...
-
+        param = model.get_parameter_or_buffer(param_name)
         param_size = int(math.prod(param.shape) * allocation_factor)
 
-        if _torch_distributed_available and torch.distributed.is_initialized():
-            generic_name = re.sub(r"\d+", "*", param_name)
-            param_size //= torch.distributed.get_world_size() if not model._tp_plan.get(generic_name, False) else 1
+        if tp_plan_regex is not None:
+            generic_name = re.sub(r"\.\d+\.", ".*.", param_name)
+            param_size //= torch.distributed.get_world_size() if tp_plan_regex.search(generic_name) else 1
```
Comment (ArthurZucker) [inline src/transformers/modeling_utils.py:5806]: does not always work, you need to get the plan's shard dim to now if weight was replicated or not!

**Review by ArthurZucker — COMMENTED** [review]
Big big PR!
